### ISSUE TITLE
Performance Optimizations?

### ISSUE THREAD

**Issue by simon-mo** [noamgat/lm-format-enforcer#28]
Hi @noamgat,

First, thank you for contributing to vLLM for the logits processors PR. I want to dive deeper into the performance of this library. Is there a way to speed things up by pre-computation of the mask per state machine similar to the outline's approach? 

(I also opened an issue asking them to leverage tokenizer prefix-tree in pre-computation pass: https://github.com/outlines-dev/outlines/issues/383)

I think there will be a lot of value in bring the performance overhead as close to 0% as possible.

**Comment by simon-mo** [comment]
Additionally, I found some overhead exists in the logits processors that moving generated mask from CPU to GPU is pretty time consuming. Doing it on very token slow things down 20-50%.

**Comment by noamgat** [comment]
Hi!
Thanks for taking interest!
I think the second point (CPU->GPU transfer) is more important, as for the first point, the calculation has to happen sometimes.
How do you see that the CPU->GPU transfer is causing a slowdown?
From my understanding, it would happen in the line
```
        mask[allowed_tokens] = 0
```
Is this the case?

**Comment by simon-mo** [comment]
Yeah I think depending on the exact configuration:
* if your mask is in CPU and need to move it to GPU that's the slowest
* if your mask is already in GPU and just resetting it with allowed_tokens, which is a list of ints that should be relatively fast.
* if your mask is in GPU and allowed_tokens is in GPU that would marginally faster because it entirely avoids memory transfers

**Comment by noamgat** [comment]
Thanks for the feedbacks everyone!
I have a day set next week to work on lm-format-enforcer, I will take care of some issues that need work. Hopefully I'll find an elegant way to work around this one as well.

**Comment by noamgat** [comment]
I commit a fix to this problem. Can you run using the master branch and check if the performance improves for your use case?

**Comment by GuyAglionby** [comment]
Thanks! I will test and get back to you

**Comment by noamgat** [comment]
Closing due to inactivity.
